### PR TITLE
Fill Order Unit tests (and other stuff)

### PR DESCRIPTION
**Pull request by dorothy-zbornak**
## Description
waaaow
### Motivation
Our `fillOrder()` tests are more integration tests than unit tests. They rely on many external token contracts, asset proxies, and complex signature validation logic to run. Because of this, it's difficult to have a high degree of confidence in correctness of the `fillOrder()` logic itself.

### Solution
To unit test `fillOrder()` directly, I've created a new  test contract [`IsolatedExchange.sol`](../blob/feature/3.0/exchange/fill-order-unit-tests/contracts/exchange/contracts/test/IsolatedExchange.sol) that:
- Overrides `_isValidOrderWithHashSignature()` to provide extremely simplified signature verification.
- Overrides `_dispatchTransferFrom()` to only log arguments, and revert if given specially crafted asset data.

With this setup, we never leave the Exchange contract and only internal functions immediately relevant to `fillOrder()` will be called. Because we log `DispatchTransferFromCalled` events, we can easily track and assert the balances of every address and asset involved in the fill *off-chain*, which is much MUCH quicker (100ms/test) than other tests that must individually query the chain for all token balances after a fill operation (1000ms/test). That's a 10x reduction!

There is also a TS wrapper class for `IsolatedExchange` called [`IsolatedExchangeWrapper`](../blob/feature/3.0/exchange/fill-order-unit-tests/contracts/exchange/test/utils/isolated_exchange_wrapper.ts) which automatically tracks balance changes for each call for you and offers other QoL improvements. It's also trivially easy to add other fill functions to this class.

In the end, we are able to provide comprehensive unit testing for `fillOrder()`, and they are wIcKeD fast, concise, and take minimal setup. See the tests in [/test/isolated_fill_order.ts](../blob/feature/3.0/exchange/fill-order-unit-tests/contracts/exchange/test/isolated_fill_order.ts).

### Other Goodies (Sorry, I got carried away)
This PR got a little out of control. There are a lot of other changes, some of which may be a little controversial.
- REMOVED `_assertValidFill()` entirely because it did nothing meaningful. This means you can pass a `0` `takerAssetFillAmount` to `fillOrder()` now.
- `SafeMath._safeDiv()` now throws a `SafeMathError(DivisionByZeroError)` if `denominator == 0`. As I was rewriting the tests for several internal/library functions, I realized it was inconsistent that many math failures threw a revert error except for `_safeDiv()`, and this function gets called a lot.
    - I've also removed a lot of the `denominator != 0` checks in several `LibMath` functions because they will be caught and thrown by `_safeDiv()` now.
- Moved `LibMath` tests out of the `exchange` package and into the `exchange-libs` package, where they belong.
    - Added a bunch of explicit tests, in addition to the combinatorial, to explicitly capture edge cases.
- REMOVED the combinatorial tests for `_addFillResults()` and `_updateFilledState()` in favor of just edge-testing. These functions are too dead simple to warrant combinatorial testing.
- Added unit tests for `_settleOrder()`.
- Added `ReferenceFunctions` exports to several packages. These provide TS implementations of useful contract functions. These have been written to be as close as possible to the actual solidity implementations (they even throw `RevertError`s) and can be used in lieu of having to call a contract to do these calculations. There are even tests for them in each package.
    - **`@0x/contracts-exchange`**:
        - `calculateFillResults()` 
    - **`@0x/contracts-exchange-libs`**:
        - `isRoundingErrorFloor()`
        - `isRoundingErrorCeil()`
        - `safeGetPartialAmountFloor()`
        - `safeGetPartialAmountCeil()`
        - `getPartialAmountFloor()`
        - `getPartialAmountCeil()`
        - `addFillResults()`
    - **`@0x/contracts-utils`**:
        - `safeDiv()`
        - `safeMul()`
        - `safeAdd()`
        - `safeSub()`
- Rewrote the combinatorial tests that are/were in `internal.ts` to be more legible/concise and to use the `ReferenceFunctions`. 
- Moved these types from `@0x/contracts-test-utils` to `@0x/types`, which make sense since these are returned by many public functions now:
    - `OrderInfo`
    - `FillResults`
    - `MatchedFillResults`
    - `BatchMatchedFillResults`
- Patched the `revertWith()` chai extensions in `@0x/dev-utils` to work with Promise-like objects:
    - This allows you to do (whereas this would have failed before):
    ```ts
        await expect(myMethod.awaitTransactionSuccessAsync()).to.revertWith(...);
    ```
- Added a bunch of constants and helpers to `@0x/contracts-test-utils`:
    - `MAX_UINT256`
    - `MAX_UINT256_ROOT`
    - `ADDRESS_LENGTH`
    - `filterLogs<T>()`
    - `filterLogsToArguments<T>()`
    - `hexRandom()`
- Hid all the combinatorial tests behind the new `describe.optional` mocha extension.
 

## Types of changes

* Bug fix (non-breaking change which fixes an issue)

* New feature (non-breaking change which adds functionality)

* Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
to the

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [x] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [x] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: contracts/exchange/contracts/test/IsolatedExchange.sol [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        );
+
+        // Fail if the first byte is 0.
+        if (assetData.length > 0 && assetData[0] == 0x00) {
+            revert("TRANSFER_FAILED");
+        }
+    }
+
+    /// @dev Overriden to simplify signature validation.
+    ///      Unfortunately, this is `view`, so it can't log arguments.
```
[the remark this answers is not in the document]
Comment (dorothy-zbornak) [reply contracts/exchange/contracts/test/IsolatedExchange.sol]: Yeah. :disappointed:

File: contracts/exchange/test/internal.ts [hunk 2]
```diff
@@ -1,561 +1,443 @@
...
+            const fillResults = {
+                makerAssetFilledAmount: ONE_ETHER.times(2),
+                takerAssetFilledAmount: ONE_ETHER.times(10),
+                makerFeePaid: ONE_ETHER.times(0.01),
+                takerFeePaid: ONE_ETHER.times(0.025),
+            };
+            const receipt = await logDecoder.getTxWithDecodedLogsAsync(
+                await testExchange.settleOrder.sendTransactionAsync(orderHash, order, takerAddress, fillResults),
+            );
+            const logs = receipt.logs as Array<
```
[the remark this answers is not in the document]
Comment (dorothy-zbornak) [reply contracts/exchange/test/internal.ts]: Wait, even though the linter complains? Seems we have to use `Array<>` for complex types (`array-type` rule).

[the remark this answers is not in the document]
Comment (abandeali1) [reply contracts/exchange/test/internal.ts]: I stand corrected!

File: contracts/exchange/tslint.json [hunk 3]
```diff
@@ -2,5 +2,8 @@
     "extends": ["@0x/tslint-config"],
     "rules": {
         "custom-no-magic-numbers": false
+    },
+    "linterOptions": {
+        "exclude": ["src/artifacts.ts"]
```
Comment (abandeali1) [inline contracts/exchange/tslint.json]: Why are we excluding this? Can we just use a `tslint:disable:next-line` for the specific line we're trying to address?

Comment (dorothy-zbornak) [reply contracts/exchange/tslint.json]: It's a generated file. For some reason it generates the imports out of order, which violates linter rules.
